drivers/sensor/vcnl4040/vcnl4040.c

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright (c) 2020 Richard Osterloh
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT vishay_vcnl4040
+
+#include "vcnl4040.h"
+#include <sys/__assert.h>
+#include <sys/byteorder.h>
+#include <sys/util.h>
+#include <logging/log.h>
+#include <stdlib.h>
+
+LOG_MODULE_REGISTER(vcnl4040, CONFIG_SENSOR_LOG_LEVEL);
+
+int vcnl4040_read(struct device *dev, uint8_t reg, uint16_t *out)
+{
+	const struct vcnl4040_config *config = dev->config;
+	struct vcnl4040_data *data = dev->data;
+
+	uint8_t buff[2] = { 0 };
+	int ret = 0;
+
+	ret = i2c_write_read(data->i2c, config->i2c_address,
+			     &reg, sizeof(reg), buff, sizeof(buff));
+
+	*out = sys_get_le16(buff);
+
+	return ret;
+}
+
+int vcnl4040_write(struct device *dev, uint8_t reg, uint16_t value)
+{
+	const struct vcnl4040_config *config = dev->config;
+	struct vcnl4040_data *data = dev->data;
+
+	struct i2c_msg msg;
+	int ret;
+	uint8_t buff[3];
+
+	sys_put_le16(value, &buff[1]);
+
+	buff[0] = reg;
+
+	msg.buf = buff;
+	msg.flags = 0;
+	msg.len = sizeof(buff);
+
+	ret = i2c_transfer(data->i2c, &msg, 1, config->i2c_address);
+
+	if (ret < 0) {
+		LOG_ERR("write block failed");
+		return -1;
+	}
+
+	return 0;
+}
+
+static int vcnl4040_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct vcnl4040_data *data = dev->data;
+	int ret = 0;
+
+	if (chan != SENSOR_CHAN_ALL) {
+		LOG_ERR("Unsupported sensor channel");
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	if (vcnl4040_read(dev, VCNL4040_REG_PS_DATA, &data->proximity)) {
+		LOG_ERR("Could not fetch proximity");
+		ret = -EIO;
+		goto exit;
+	}
+#ifdef CONFIG_VCNL4040_ENABLE_ALS
+	const struct vcnl4040_config *config = dev->config;
+
+	if (vcnl4040_read(dev, VCNL4040_REG_ALS_DATA, &data->light)) {
+		LOG_ERR("Could not fetch ambient light");
+		ret = -EIO;
+		goto exit;
+	}
+	/*
+	 * scale the lux depending on the value of the integration time
+	 * see page 8 of the VCNL4040 application note:
+	 * https://www.vishay.com/docs/84307/designingvcnl4040.pdf
+	 */
+	data->light = (data->light * (0.1 / (1 << config->als_it)));
```

Comment:
`(10 * data->light) >> config->als_it` would avoid the need to pull in floating point.

Reply:
Compiler warns that this statement does nothing so I've left this change out

Reply:
I meant using that as the right hand side of the assignment.

Reply:
Ok, changed it for you anyway

Reply:
Thanks.  Please do test it, what I proposed is supposed to be equivalent in integer math but I may have been wrong.

Reply:
Sorry to be a pain. I've had another look at the datasheet and tested and I think it needs to be a float. Would you mind having another look. My tests show that the maximum lux values for the particular sensitivities hold true.

Reply:
Testing should usually win over theory, and yes I was wrong: it should have been:

    (data->light >> config_als_it) / 10

Multiplying by 0.1 is equivalent to dividing by 10.

But if you want to just leave it as floating point that's fine.